### PR TITLE
Introduce reset function for the navigator

### PR DESCRIPTION
**Pull request by kmadsen**
## Description

Opening this as a draft while working on examples that test out the function.

We are looking to add the ability to teleport the navigator to different locations. For example, simulate a route in new york and then jump to tokyo and simulate a new route - all while keeping the same SDK setup.

A more practical case, would be to simulate a route in your city. And then jump back to the beginning to drive it for real.

- [ ] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

## Testing

Notice that without the reset function, the navigator will not reliably reset to the new location selected.

| without reset function | with reset function |
| -- | -- |
| output | output |

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt [hunk 1]
```diff
@@ -86,6 +87,29 @@ class ReplayHistoryActivity : AppCompatActivity() {
...
+        }
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun ReplayNavigationContext.handleHistoryFileSelected() {
+        loadNavigationJob = CoroutineScope(Dispatchers.Main).launch {
+            val events = loadReplayHistory()
+            mapboxReplayer.clearEvents()
+            mapboxReplayer.pushEvents(events)
+            mapboxNavigation.resetTripSession()
```
Comment (kmadsen) [inline examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:104]: this line is what causes the difference between the videos in the PR description.

The navigator needs to be reset, and then it can start from a new location.

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt [hunk 2]
```diff
@@ -64,15 +65,15 @@ class ReplayHistoryActivity : AppCompatActivity() {
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_replay_history_layout)
         mapView.onCreate(savedInstanceState)
 
         selectHistoryButton.setOnClickListener {
             val activityIntent = Intent(this, HistoryFilesActivity::class.java)
-            startActivity(activityIntent)
-            finish()
+            startActivityForResult(activityIntent, HistoryFilesActivity.REQUEST_CODE)
```
Comment (kmadsen) [inline examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:76]: Updating the example to use `startActivityForResult` to showcase the `resetTripSession` function.

When we start an activity for result, the current activity is never destroyed. Making it so, when the onActivityResult comes back it will be able to update the state of MapboxNavigation without completely restoring it.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 3]
```diff
@@ -262,6 +262,14 @@ class MapboxNavigation(
         tripSession.stop()
     }
 
+    /**
+     * Clears the location and route for the navigator.
+     */
+    fun resetTripSession() {
+        tripSession.route = null
+        navigator.reset()
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:270]: Digging up some history as we had a similar function in the past.

Introduced a reset function, that would recreate the navigator https://github.com/mapbox/mapbox-navigation-android/pull/2850
That caused a regression in the examples https://github.com/mapbox/mapbox-navigation-android/issues/2769
Fixed the regression with https://github.com/mapbox/mapbox-navigation-android/pull/2905

Notice, the issue was because the reset function was called inside [MapboxTripSession.reset()](https://github.com/mapbox/mapbox-navigation-android/pull/2905/files/d48bb7ec1c23eeab1530de594bd978c67ec660a7#diff-f1923dcc227762ccf3eb5fcf1cab79bc5da009704c56901e0c78b6d257b5c1d6L187)

[But we're not calling it here](https://github.com/mapbox/mapbox-navigation-android/blob/aa333725a94a0cbd2edf7c79fd08879b8c5a35ad/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L205-L214)

Worth noting that, our MapboxTripSession private reset function is very different from the Nav Native's resetRideSession.

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:270]: I've also tested the repro steps included in the issue from before: https://github.com/mapbox/mapbox-navigation-android/issues/2769#issuecomment-616738902

And this change is looking good. It's because we don't want to add the Navigator.resetRideSession call to the MapboxTripSession.reset() function

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 4]
```diff
@@ -262,6 +262,14 @@ class MapboxNavigation(
         tripSession.stop()
     }
 
+    /**
+     * Clears the location and route for the navigator.
+     */
+    fun resetTripSession() {
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:268]: Can we add a couple of tests around this new API?

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt [hunk 5]
```diff
@@ -64,6 +64,10 @@ object MapboxNativeNavigatorImpl : MapboxNativeNavigator {
         return this
     }
 
+    override fun reset() {
+        navigator?.resetRideSession()
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigatorImpl.kt:68]: Do we want to crash if `reset` is called before `create`? We're using `!!` across the rest of the APIs instead of `?.` 🤔

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt [hunk 6]
```diff
@@ -34,6 +34,13 @@ interface MapboxNativeNavigator {
         tilesConfig: TilesConfig
     ): MapboxNativeNavigator
 
+    /**
+     * Reset the navigator state with the same configuration. The location becomes unknown,
+     * but the [NavigatorConfig] stays the same. This can be used to transport the
```
Comment (Guardiola31337) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt:39]: Should we also some of this to `MapboxNavigation#resetTripSession` KDocs?

Comment (Guardiola31337) [reply libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt:39]: Yeah, I meant adding some clarifications here https://github.com/mapbox/mapbox-navigation-android/blob/d00ec6150cde9352c7b352d068b204dd89c0cefb/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L265-L267 For example `This can be used to transport the...`

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt (hunk 4 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:268]: Test are failing in CI https://app.circleci.com/pipelines/github/mapbox/mapbox-navigation-android/9325/workflows/ab41e4ad-7d22-4571-b82b-ce0f136001ec/jobs/34959

```
> Task :libnavigation-core:compileDebugUnitTestKotlin FAILED
e: /root/code/libnavigation-core/src/test/java/com/******/navigation/core/MapboxNavigationTest.kt: (504, 53): Unresolved reference: Collections
```

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt (hunk 6 above)
Comment (kmadsen) [reply libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt:39]: OH yeah I misinterpreted. Thanks for clarifying 😅

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt [hunk 7]
```diff
@@ -289,10 +289,10 @@ class MapboxNavigationTest {
     }
 
     @Test
-    fun onDestroySetsRouteToNullInTripSession() {
+    fun onDestroySetsRoutesToEmptyInTripSession() {
         mapboxNavigation.onDestroy()
 
-        verify(exactly = 1) { tripSession.route = null }
+        verify(exactly = 1) { directionsSession.routes = emptyList() }
```
Comment (kmadsen) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationTest.kt:295]: directionsSession is supposed to update tripsSession.route

so our real source of truth is the directionsSession

it's a bit more complicated to wire up tests to have this assumption built in

File: libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt [hunk 8]
```diff
@@ -34,6 +34,13 @@ interface MapboxNativeNavigator {
...
+     * navigator to a new location.
+     */
+    fun reset()
```
Comment (LukasPaczos) [inline libnavigator/src/main/java/com/mapbox/navigation/navigator/internal/MapboxNativeNavigator.kt:42]:
```suggestion
    fun resetRideSession()
```

Let's call it the same as the native API.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 9]
```diff
@@ -262,6 +262,16 @@ class MapboxNavigation(
         tripSession.stop()
     }
 
+    /**
+     * Reset the session with the same configuration. The location becomes unknown,
+     * but the [NavigationOptions] stay the same. This can be used to transport the
+     * navigator to a new location.
+     */
+    fun resetTripSession() {
+        directionsSession.routes = emptyList()
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:271]: Do we need to reset the route as well? I'd expect this method to only reset the Navigator's bias (we can work on the method naming as well) and nothing else.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:271]: While reviewing this noticed that we were not resetting the `directionsSession.routes` as part of `onDestroy` which allows you to do

```kotlin
mapboxNavigation.onDestroy()
mapboxNavigation.getRoutes()
```

We considered adding it here which is called when in `onDestroy`. If we opt to remove it we should add it after `directionsSession.unregisterAllRoutesObservers()` when in `onDestroy` (we can do it in a separate PR though as it doesn't directly relate to the changes added as part of this PR).

cc @kmadsen

Comment (kmadsen) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:271]: We agreed in comments that native was going to reset the route, and the SDK callers would need to set route again if they wanted to keep it. But that is actually, not true.

``` c++
void NavigatorImpl::resetRideSession() {
  BuildRideSessionImpl();
  if (routes_ && ride_session_) {
    try {
      ride_session_->SetRoutes(routes_);
    } catch (const std::exception& ex) {
      LOG_ERROR(std::string{"SetRoute error: "} + ex.what());
    }
  }
```

cc: @mskurydin @SiarheiFedartsou @averkhaturau

I think @LukasPaczos is right, we don't want to reset the route inside this `resetTripSession` function. So am moving the route reset back to `onDestroy`

**Review by LukasPaczos — APPROVED** [review]
LGTM

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt [hunk 10]
```diff
@@ -342,14 +351,8 @@ class MapboxNavigation(
...
         tripSession.unregisterAllEHorizonObservers()
-        tripSession.route = null
-
-        // TODO replace this with a destroy when nav-native has a destructor
-        navigator.create(
-            navigationOptions.deviceProfile,
-            navigatorConfig,
-            createTilesConfig()
-        )
+        directionsSession.routes = emptyList()
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt:354]: 👍

**Review by Guardiola31337 — APPROVED** [review]

**Comment by kmadsen** [comment]
robo tests [ran out of memory](https://console.firebase.google.com/u/0/project/mapbox-navigation-android/testlab/histories/bh.2e74b6b076d6fa53/matrices/5126404865723193639/executions/bs.e0187974ad3ee6c0/issues). wonder if there is a memory leak

```

java.lang.OutOfMemoryError: Could not allocate JNI Env
     FATAL EXCEPTION: main
Process: com.mapbox.navigation.examples, PID: 13039
java.lang.OutOfMemoryError: Could not allocate JNI Env
	at java.lang.Thread.nativeCreate(Native Method)
	at java.lang.Thread.start(Thread.java:1063)
	at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:921)
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1339)
	at f.f0$a.m()
	at f.s.h()
	at f.s.a()
	at f.f0.s()
	at h.n.E()
	at h.i$b.E()
	at d.b.c.a.a.i.c()
	at d.b.f.e.a.a.a.getRoute()
	at com.mapbox.navigation.route.internal.hybrid.MapboxHybridRouter$d.e()
	at com.mapbox.navigation.route.internal.hybrid.MapboxHybridRouter$d.getRoute()
	at com.mapbox.navigation.route.internal.hybrid.MapboxHybridRouter.getRoute()
	at d.b.f.b.l.a.b.a()
	at d.b.f.b.p.a.i()
	at d.b.f.b.p.a.b()
	at d.b.f.b.a.G()
	at d.b.f.b.a.c()
	at d.b.f.b.a$d.t()
	at d.b.f.b.t.a.g.V()
	at d.b.f.b.t.a.g.J()
	at d.b.f.b.t.a.g$f.l()
	at e.v.k.a.a.i()
	at kotlinx.coroutines.x0.run()
	at android.os.Handler.handleCallback(Handler.java:746)
	at android.os.Handler.dispatchMessage(Handler.java:95)
	at android.os.Looper.loop(Looper.java:148)
	at android.app.ActivityThread.main(ActivityThread.java:5443)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:728)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:618)
```

**Comment by kmadsen** [comment]
Still a pain to deobfuscate stack traces, [here's the instructions](https://github.com/mapbox/mapbox-navigation-android/issues/3562#issuecomment-707265067) 
[mapping.txt](https://github.com/mapbox/mapbox-navigation-android/files/5477792/mapping.txt)

But the actual crash is coming from the requesting a route from HTTP
```
com.mapbox.navigation.route.internal.offboard.MapboxOffboardRouter$getRoute$1 -> d.b.f.e.a.a.a$b:
    com.mapbox.navigation.base.route.Router$Callback $callback -> a
    void onResponse(retrofit2.Call,retrofit2.Response) -> a
    void onFailure(retrofit2.Call,java.lang.Throwable) -> b
```

```
okhttp3.RealCall$AsyncCall -> f.f0$a:
    okhttp3.Callback responseCallback -> f
    java.util.concurrent.atomic.AtomicInteger callsPerHost -> g
    okhttp3.RealCall this$0 -> h
    void execute() -> k
    java.util.concurrent.atomic.AtomicInteger callsPerHost() -> l
    void executeOn(java.util.concurrent.ExecutorService) -> m
```

Syncing with @averkhaturau. We don't believe this is due to this change at this time. The crash is coming from the network call rather than directly from nav-native. 

I'll run one more test before merging

**Comment by kmadsen** [comment]
tl;dr - I'm not finding any reason to believe this pull request is introducing an issue

After running some tests locally, not able to identify a leak from requesting routes. But the "Time to request a route" is very slow - I was able to reproduce this on master and it was possibly due to my own network quality.

Screen Shot 2020-11-02 at 12 25 00 PM

output
